js/modules/types/contact.js

```diff
@@ -0,0 +1,142 @@
+const { omit, compact, map } = require('lodash');
+
+const { toLogFormat } = require('./errors');
+const { SignalService } = require('../../../ts/protobuf');
+
+const DEFAULT_PHONE_TYPE = SignalService.DataMessage.Contact.Phone.Type.HOME;
+const DEFAULT_EMAIL_TYPE = SignalService.DataMessage.Contact.Email.Type.HOME;
+const DEFAULT_ADDRESS_TYPE =
+  SignalService.DataMessage.Contact.PostalAddress.Type.HOME;
+
+exports.parseAndWriteContactAvatar = upgradeAttachment => async (
+  contact,
+  context = {}
+) => {
+  const { message } = context;
+  const { avatar } = contact;
+  const contactWithUpdatedAvatar =
+    avatar && avatar.avatar
+      ? Object.assign({}, contact, {
+          avatar: Object.assign({}, avatar, {
+            avatar: await upgradeAttachment(avatar.avatar, context),
+          }),
+        })
+      : omit(contact, ['avatar']);
+
+  // eliminates empty numbers, emails, and addresses; adds type if not provided
+  const contactWithCleanedElements = parseContact(contactWithUpdatedAvatar);
+
+  const error = exports._validateContact(contactWithCleanedElements, {
+    messageId: idForLogging(message),
+  });
+  if (error) {
+    console.log(
+      'Contact.parseAndWriteContactAvatar: contact was malformed.',
+      toLogFormat(error)
+    );
+  }
+
+  return contactWithCleanedElements;
+};
+
+function parseContact(contact) {
+  return Object.assign(
+    {},
+    omit(contact, ['avatar', 'number', 'email', 'address']),
+    parseAvatar(contact.avatar),
+    createArrayKey('number', compact(map(contact.number, cleanPhoneItem))),
+    createArrayKey('email', compact(map(contact.email, cleanEmailItem))),
+    createArrayKey('address', compact(map(contact.address, cleanAddress)))
+  );
+}
+
+function idForLogging(message) {
+  return `${message.source}.${message.sourceDevice} ${message.sent_at}`;
+}
+
+exports._validateContact = (contact, options = {}) => {
+  const { messageId } = options;
+  const { name, number, email, address, organization } = contact;
+
+  if ((!name || !name.displayName) && !organization) {
+    return new Error(
+      `Message ${messageId}: Contact had neither 'displayName' nor 'organization'`
+    );
+  }
+
+  if (
+    (!number || !number.length) &&
+    (!email || !email.length) &&
+    (!address || !address.length)
+  ) {
+    return new Error(
+      `Message ${messageId}: Contact had no included numbers, email or addresses`
+    );
+  }
+
+  return null;
+};
```

Comment:
Thanks for doing the refactor! This was inspired by some reading about [`purescript-validation`](https://pursuit.purescript.org/packages/purescript-validation/3.2.0/docs/Data.Validation.Semigroup#t:V). Sadly, I can’t find the original blog post anymore but this one comes close: https://medium.com/blacklane-engineering/pure-functional-validation-64a7885d22ac It’s a bit dense but I am happy to go over it if you’re interested 😄